design/src/rfcs/rfc0023_request_id.md

```diff
@@ -0,0 +1,105 @@
+[2](https://cwiki.apache.org/confluence/display/WW/S2-045)).
+Having the client send the value bring the following challenges:
+* The client could repeatedly send the same RequestID
+* The client could send no RequestID
+* The client could send a malformed or malicious RequestID
+
+To minimise the attack surface and provide a uniform experience to customers, servers must generate the value.
+Services are free to read the ID sent by clients in HTTP headers.
+
+RequestIDs are not to be used by multiple services, but only within a service.
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+The proposal is to implement a `RequestId` type and make it available to middleware and business logic handlers, through [FromParts](../server/from-parts.md) and as a `Service`.
+
+1. Implementing `FromParts` for `Extension<RequestId>` gives customers the ability to write their handlers:
+
+```rust
+pub async fn handler(
+    input: input::Input,
+    request_id: Extension<RequestId>,
+) -> ...
+```
+
+`RequestId` will be injected into the extensions by a layer.
+This layer can also be used to open a span that will log the request ID: subsequent logs will be in the scope of that span.
+
+2. RequestId format:
+
+Common formats for RequestIDs are:
+
+* UUID: a random string, represented in hex, of 128 bits from IETF RFC 4122: `7c038a43-e499-4162-8e70-2d4d38595930`
+* The hash of a sequence such as `date+thread+server`: `734678902ea938783a7200d7b2c0b487`
+* A verbose description: `current_ms+hostname+increasing_id`
+
+For privacy reasons, any format that provides service details should be avoided. A random string is preferred.
+The proposed format is to use UUID, version 4.
+
+AWS Lambda sends a RequestID in the request context; there is no need to modify it.
+There should be a `Service` for Lambda users that extracts the value and injects it as for non-Lambda other requests.
+
+A `Service` that inserts a RequestId in the extensions:
+```rust
+impl<R, S> Service<http::Request<R>> for RequestIdProvider<S>
+where
+    S: Service<http::Request<R>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: http::Request<R>) -> Self::Future {
+        req.extensions_mut().insert(RequestId::new());
+        self.inner.call(req)
+    }
+}
+```
+
+A `Service` that inserts a RequestId in the extensions:
+```rust
+fn call(&mut self, mut req: http::Request<R>) -> Self::Future {
+    let req_id = RequestId::new();
+    let span = span!(request_id = %req_id);
+    request.extensions_mut().insert(req_id);
+    self.inner.call(request).instrument(span)
+}
+```
```

Comment:
I'd advice against coupling the insertion of `RequestId` into the extensions map with the creation of a root span.
Most services will want to customise the root span and we don't want to introduce an additional unnecessary span in their telemetry data.

We could provide a basic `TelemetryService` that could be used as default and takes care of injecting the `RequestId`, but it should be a separate service.

Reply:
Gotcha - I read this snippet as "this is what the service will look like" instead of "this is what users could do with it". 
If the latter, I'd expect them to retrieve an already-generated `RequestId` from the extensions rather than doing the work of creating and storing it.

Reply:
I'd lean against that - if they end up using both layers by mistake, things will get confusing since they will get two different ids generated and used in different parts of the app.
That won't happen if `RequestIdService` and `TelemetryService` are orthogonal (i.e. you can use the first or both, but the latter requires the first to get a valid `RequestId`).

Reply:
The downside of vending them orthogonal is that `TelemetryService` is an additional and fallible type map lookup. There's also a question here of whether or not a `span!` with a lone `RequestId` is a problem?

If we do decide to do it we should perhaps look to extend `aws-smithy-http-server::instrumentation::service::InstrumentOperation`.

Reply:
> There's also a question here of whether or not a span! with a lone RequestId is a problem?

I'd say that it depends on the span level. At the `info` level, I think it's a problem - it's most likely not going to be filtered out in production and it's going to add 1 or 2 telemetry data points for every request - that can be costly and add to the overall telemetry noise if it's not providing enough information on its own.

> The downside of vending them orthogonal is that TelemetryService is an additional and fallible type map lookup. 
> If we do decide to do it we should perhaps look to extend aws-smithy-http-server::instrumentation::service::InstrumentOperation.

I think that it's OK to couple generation and logging if we allow for the insertion of custom properties into the span - then the user can benefit from out default plus whatever they care about (e.g. something like [`root_span!`](https://docs.rs/tracing-actix-web/latest/tracing_actix_web/macro.root_span.html) is one way of doing it).
`InstrumentOperation` seems like a good  spot to build on top of.

Reply:
There’s a reasonable chunk of design work to be done around the `InstrumentOperation` API, at the time of writing,  just wraps the `http::Request` and `http::Response` parts with sensitive wrappers, opens a request span and emits an response event - no user configuration possible.

Perhaps we want it closer to what `TraceLayer` does, it’s unclear to me whether we reuse `TraceLayer` or not because the closures provided to each `TraceLayer` are ~ `|request: http::Request| Span/Event` / `|response: http::Response| Span/Event` / etc and we don’t really want to provide the customer with the entire `http::Request` - we instead want to provide them with the `http::Request`/`http::Response` parts in wrappers which redact sensitive information.

Inserting(?) and recording a `RequestId` is a similar level of configuration which could be added to the `InstrumentOperation`.